File: src/Medium/LinkListTest/MyCircularQueue.java
package Medium.LinkListTest;

public class MyCircularQueue {


    private int[] arr;
    private int rear;
    private int front;
    private int size;
    private boolean tag;
    /** Initialize your data structure here. Set the size of the queue to be k. */
    public MyCircularQueue(int k) {
        size=k;
        arr=new int[k];
        front=0;
        rear=0;
        tag=false;
    }

    /** Insert an element into the circular queue. Return true if the operation is successful. */
    public boolean enQueue(int value) {
        if(isFull()){
            return false;
        }else {
            arr[rear]=value;
            rear=(rear+1)%size;
            tag=true;
            return true;
        }
    }

    /** Delete an element from the circular queue. Return true if the operation is successful. */
    public boolean deQueue() {
        if(isEmpty()) return false;
        front=(front+1)%size;
        tag=false;
        return true;
    }

    /** Get the front item from the queue. */
    public int Front() {
        return isEmpty()?-1:arr[front];
    }

    /** Get the last item from the queue. */
    public int Rear() {
        return isEmpty()?-1:arr[(rear-1+size)%size];
    }

    /** Checks whether the circular queue is empty or not. */
    public boolean isEmpty() {
        return rear==front&&!tag;
    }

    /** Checks whether the circular queue is full or not. */
    public boolean isFull()  {
        return rear==front&&tag;
    }

    public static void main(String[] args) {

    }
}

File: src/Medium/DesignTest/MyCircularQueue.py
class MyCircularQueue:

    def __init__(self, k: int):
        self.q = [0] * (k + 1)
        self.f = 0
        self.r = 0
        self.n = k + 1

    def enQueue(self, value: int) -> bool:
        if self.isFull():
            return False
        self.q[self.r] = value
        self.r = (self.r + 1) % self.n
        return True

    def deQueue(self) -> bool:
        if self.isEmpty():
            return False
        self.f = (self.f + 1) % self.n
        return True

    def Front(self) -> int:
        return -1 if self.isEmpty() else self.q[self.f]

    def Rear(self) -> int:
        return -1 if self.isEmpty() else self.q[(self.r - 1 + self.n) % self.n]

    def isEmpty(self) -> bool:
        return self.f == self.r

    def isFull(self) -> bool:
        return (self.r + 1) % self.n == self.f


if __name__ == '__main__':
    # obj = MyCircularQueue(k)
    # param_1 = obj.enQueue(value)
    # param_2 = obj.deQueue()
    # param_3 = obj.Front()
    # param_4 = obj.Rear()
    # param_5 = obj.isEmpty()
    # param_6 = obj.isFull()
    pass
